lib/fs.js

```diff
@@ -1404,34 +1404,64 @@ function mkdirSync(path, options) {
   }
 }
 
-// TODO(Ethan-Arrowood): Make this iterative too
-function readdirSyncRecursive(path, origPath, options) {
-  nullCheck(path, 'path', true);
-  const ctx = { path };
-  const result = binding.readdir(pathModule.toNamespacedPath(path),
-                                 options.encoding, !!options.withFileTypes, undefined, ctx);
-  handleErrorFromBinding(ctx);
-  return options.withFileTypes ?
-    getDirents(path, result).flatMap((dirent) => {
-      return [
-        dirent,
-        ...(dirent.isDirectory() ?
-          readdirSyncRecursive(
-            pathModule.join(path, dirent.name),
-            origPath,
-            options,
-          ) : []),
-      ];
-    }) :
-    result.flatMap((ent) => {
-      const innerPath = pathModule.join(path, ent);
-      const relativePath = pathModule.relative(origPath, innerPath);
-      const stat = binding.internalModuleStat(innerPath);
-      return [
-        relativePath,
-        ...(stat === 1 ? readdirSyncRecursive(innerPath, origPath, options) : []),
-      ];
-    });
+/**
+ * An iterative algorithm for reading the entire contents of the `basePath` directory.
+ * This function does not validate `basePath` as a directory. It is passed directly to
+ * `binding.readdir` after a `nullCheck`.
+ * @param {string} basePath
+ * @param {{ encoding: string, withFileTypes: boolean }} options
+ * @returns {Array<string> | Array<Dirent>}
+ */
+function readdirSyncRecursive(basePath, options) {
+  nullCheck(basePath, 'path', true);
+
+  const results = [];
+  const opsQueue = [];
+
+  function _read(_path) {
+    const ctx = { _path };
+    const result = binding.readdir(
+      pathModule.toNamespacedPath(_path),
+      options.encoding,
+      !!options.withFileTypes,
+      undefined,
+      ctx,
+    );
+    handleErrorFromBinding(ctx);
+
+    if (options.withFileTypes) {
+      const dirents = getDirents(_path, result);
+      for (let i = 0; i < dirents.length; i++) {
+        const dirent = dirents[i];
+        ArrayPrototypePush(results, dirent);
+        if (dirent.isDirectory()) {
+          ArrayPrototypePush(opsQueue, curryRead(pathModule.join(dirent.path, dirent.name)));
+        }
+      }
+    } else {
+      for (let i = 0; i < result.length; i++) {
+        const ent = result[i];
```

Comment:
Variable names are quite ambiguous. Can I recommend choosing a more verbal and descriptive names?